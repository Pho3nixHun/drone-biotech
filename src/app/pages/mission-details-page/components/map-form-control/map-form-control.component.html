<div
    appMapInitializer
    [signal]="mapSignal"
    [mapOptions]="mapOptions"
    style="height: 500px"
></div>

<app-map-area-select-form-control
    [mapRef]="mapSignal()"
    [formControl]="mapForm.controls.targetArea"
/>
<app-map-point-select-form-control
    [mapRef]="mapSignal()"
    [formControl]="mapForm.controls.entryPoint"
/>
