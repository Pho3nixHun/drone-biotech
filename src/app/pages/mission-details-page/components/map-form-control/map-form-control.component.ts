import {
    Component,
    computed,
    effect,
    forwardRef,
    inject,
    signal,
} from '@angular/core';
import {
    ControlValueAccessor,
    FormBuilder,
    NG_VALUE_ACCESSOR,
    ReactiveFormsModule,
} from '@angular/forms';
import { Coordinates } from '@stores/location/location.model';
import { MapAreaSelectFormControlComponent } from './components/map-area-select-form-control/map-area-select-form-control.component';
import { MapPointSelectFormControlComponent } from './components/map-point-select-form-control/map-point-select-form-control.component';
import { noop } from 'rxjs';
import { getCenter } from 'geolib';
import { MapControl } from '../../mission-details-page.model';
import { MapInitializerDirective } from '@directives/map-initializer/map-initializer.directive';
import { MAP_OPTIONS } from 'src/app/shared/providers/google-maps-provider';
import { toSignal } from '@angular/core/rxjs-interop';

/**
 * MapFormControlComponent
 *
 * Type: Container
 *
 * Scope:
 * - Renders a form control that handle the relevant information about the mission like the target area, entry point. It creates a Google Map to visualize the data.
 * - Passes relevant data, like config, texts.
 *
 * Out-of-Scope:
 * - Does not handle the internal logic or styling of other components.
 * - Not responsible for the detailed presentation logic.
 *
 * Purpose (optional):
 * - To serve as a smart container component that integrates business logic, including presentation, to create a cohesive user interface.
 */
@Component({
    selector: 'app-map-form-control',
    imports: [
        MapAreaSelectFormControlComponent,
        MapPointSelectFormControlComponent,
        MapInitializerDirective,
        ReactiveFormsModule,
    ],
    providers: [
        {
            provide: NG_VALUE_ACCESSOR,
            useExisting: forwardRef(() => MapFormControlComponent),
            multi: true,
        },
    ],
    templateUrl: './map-form-control.component.html',
})
export class MapFormControlComponent implements ControlValueAccessor {
    private readonly fb = inject(FormBuilder);
    protected readonly mapOptions = inject(MAP_OPTIONS);
    protected readonly mapSignal = signal<google.maps.Map | null>(null);

    protected readonly mapForm = this.fb.group({
        targetArea: this.fb.control<Coordinates[] | null>(null),
        entryPoint: this.fb.control<Coordinates | null>(null),
    });

    private readonly targetArea = toSignal(
        this.mapForm.controls.targetArea.valueChanges
    );

    private readonly mapCenter = computed<Coordinates | null>(() => {
        const area = this.targetArea();
        if (!area) return null;
        const center = getCenter(area);
        return center
            ? {
                  lat: center.latitude,
                  lng: center.longitude,
              }
            : null;
    });

    private readonly setMapCenterEffect = effect(
        () => {
            const map = this.mapSignal();
            const center = this.mapCenter();
            if (!map || !center) return;
            map.setCenter(center);
            this.setMapCenterEffect.destroy();
        },
        { manualCleanup: true }
    );

    public writeValue(value: MapControl) {
        this.mapForm.setValue(value);
    }

    private onChange: (value: MapControl) => void = () => noop;

    private onTouched: () => void = () => noop;

    public registerOnChange(fn: () => void): void {
        this.onChange = fn;
    }

    public registerOnTouched(fn: () => void): void {
        this.onTouched = fn;
    }
}
