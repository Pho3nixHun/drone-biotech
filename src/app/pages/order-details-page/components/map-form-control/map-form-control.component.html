@let vM = vm();

<div
    appMapInitializer
    [signal]="mapSignal"
    [mapOptions]="mapOptions"
    style="height: 500px"
    class="mb-4"
></div>

@if (mapSignal(); as map) {
    <app-area-form-control
        [map]="map"
        [formControl]="mapFormGroup.controls.targetAreas"
    />
}

@if (missionStats(); as stats) {
    <app-info-item-list
        class="*:flex-row *:space-y-0 *:justify-between"
        *transloco="let t"
    >
        <app-info-item class="*:flex-col">
            <dt class="text-gray-700">{{ t(vM.totalMissionsLabelKey) }}</dt>
            <dd class="font-semibold">
                {{
                    t(vM.totalMissionsValueKey, { count: stats.totalMissions })
                }}
            </dd>
        </app-info-item>
        <app-info-item class="*:flex-col">
            <dt class="text-gray-700">{{ t(vM.completedMissionsLabelKey) }}</dt>
            <dd class="font-semibold">
                {{
                    t(vM.completedMissionsValueKey, {
                        count: stats.completedMissions,
                    })
                }}
            </dd>
        </app-info-item>
        <app-info-item class="*:flex-col">
            <dt class="text-gray-700">{{ t(vM.remainingMissionsLabelKey) }}</dt>
            <dd class="font-semibold">
                {{
                    t(vM.remainingMissionsValueKey, {
                        count: stats.remainingMissions,
                    })
                }}
            </dd>
        </app-info-item>
    </app-info-item-list>
}
