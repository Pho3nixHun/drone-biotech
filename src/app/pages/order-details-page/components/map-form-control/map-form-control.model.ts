export interface MissionStats {
    totalMissions: number;
    completedMissions: number;
    remainingMissions: number;
}

export interface MapFormControlVM {
    totalMissionsLabelKey: string;
    totalMissionsValueKey: string;
    completedMissionsLabelKey: string;
    completedMissionsValueKey: string;
    remainingMissionsLabelKey: string;
    remainingMissionsValueKey: string;
}
