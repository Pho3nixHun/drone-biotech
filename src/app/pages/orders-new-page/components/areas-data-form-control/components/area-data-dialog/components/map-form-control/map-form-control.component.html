<div class="flex flex-col space-y-2">
    <app-map-search-input-form-control
        [defaultCenter]="vm().defaultCenter"
        [vm]="vm().mapSearchInputFormControlVM"
        (changeBounds)="setMapBounds($event)"
    />

    <div appElementRef [signal]="divSignal" style="height: 600px"></div>

    <app-map-area-select-form-control
        [mapRef]="mapSignal()"
        [vm]="vm().mapAreaSelectFormControlVM"
        [formControl]="mapForm.controls.targetArea"
    />
    <app-map-point-select-form-control
        [mapRef]="mapSignal()"
        [vm]="vm().mapPointSelectFormControlVM"
        [formControl]="mapForm.controls.entryPoint"
    />
</div>
