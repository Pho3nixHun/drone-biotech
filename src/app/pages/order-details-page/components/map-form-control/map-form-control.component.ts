import {
    Component,
    computed,
    effect,
    forwardRef,
    inject,
    input,
    signal,
} from '@angular/core';
import {
    ControlValueAccessor,
    FormBuilder,
    NG_VALUE_ACCESSOR,
    ReactiveFormsModule,
} from '@angular/forms';
import { Coordinates } from '@stores/location/location.model';
import { noop } from 'rxjs';
import { getCenter } from 'geolib';
import { AreaFormControlComponent } from './components/area-form-control/area-form-control.component';
import { TargetArea } from '../../order-details-page.model';
import { MAP_OPTIONS } from 'src/app/shared/providers/google-maps-provider';
import { MapFormControlVM, MissionStats } from './map-form-control.model';
import { toSignal } from '@angular/core/rxjs-interop';
import { TranslocoModule } from '@jsverse/transloco';
import { InfoItemListComponent } from '@components/info-panel/components/info-item-list/info-item-list.component';
import { InfoItemComponent } from '@components/info-panel/components/info-item-list/components/info-item/info-item.component';
import { MapInitializerDirective } from '@directives/map-initializer/map-initializer.directive';

@Component({
    selector: 'app-map-form-control',
    imports: [
        AreaFormControlComponent,
        ReactiveFormsModule,
        TranslocoModule,
        InfoItemListComponent,
        InfoItemComponent,
        MapInitializerDirective,
    ],
    providers: [
        {
            provide: NG_VALUE_ACCESSOR,
            useExisting: forwardRef(() => MapFormControlComponent),
            multi: true,
        },
    ],
    templateUrl: './map-form-control.component.html',
})
export class MapFormControlComponent implements ControlValueAccessor {
    private readonly fb = inject(FormBuilder);
    protected readonly mapOptions = inject(MAP_OPTIONS);

    public vm = input.required<MapFormControlVM>();

    protected readonly mapSignal = signal<google.maps.Map | null>(null);

    protected readonly mapFormGroup = this.fb.group({
        targetAreas: this.fb.control<TargetArea[] | null>(null),
    });

    private readonly targetAreas = toSignal(
        this.mapFormGroup.controls.targetAreas.valueChanges
    );

    protected readonly missionStats = computed<MissionStats | null>(() => {
        const totalMissions = this.targetAreas();
        if (!totalMissions) return null;

        const remainingMissions = totalMissions.filter(
            (area) => area.type === 'active'
        );

        return {
            totalMissions: totalMissions.length,
            remainingMissions: remainingMissions.length,
            completedMissions: totalMissions.length - remainingMissions.length,
        };
    });

    protected readonly mapCenter = computed<Coordinates | null>(() => {
        const areas = this.targetAreas();
        if (!areas) return null;

        const center = getCenter(areas.flatMap((area) => area.coordinates));
        return center
            ? {
                  lat: center.latitude,
                  lng: center.longitude,
              }
            : null;
    });

    protected readonly setMapCenterEffect = effect(() => {
        const map = this.mapSignal();
        const center = this.mapCenter();
        if (!map || !center) return;
        map.setCenter(center);
        this.setMapCenterEffect.destroy();
    });

    public writeValue(targetAreas: TargetArea[]) {
        this.mapFormGroup.setValue({ targetAreas });
    }

    private onChange: (value: TargetArea[]) => void = () => noop;

    private onTouched: () => void = () => noop;

    public registerOnChange(fn: () => void): void {
        this.onChange = fn;
    }

    public registerOnTouched(fn: () => void): void {
        this.onTouched = fn;
    }
}
