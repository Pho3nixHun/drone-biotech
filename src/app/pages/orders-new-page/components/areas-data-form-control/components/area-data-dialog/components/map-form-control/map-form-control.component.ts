/* eslint-disable @typescript-eslint/no-empty-function */
import {
    Component,
    computed,
    effect,
    ElementRef,
    forwardRef,
    inject,
    input,
    Signal,
    signal,
} from '@angular/core';
import {
    ControlValueAccessor,
    FormBuilder,
    NG_VALUE_ACCESSOR,
    ReactiveFormsModule,
    Validators,
} from '@angular/forms';
import { MapFormControlVM } from './map-form-control.model';
import { Coordinates } from '@stores/location/location.model';
import { MapAreaSelectFormControlComponent } from './components/map-area-select-form-control/map-area-select-form-control.component';
import { MapPointSelectFormControlComponent } from './components/map-point-select-form-control/map-point-select-form-control.component';
import { MapSearchInputFormControlComponent } from './components/map-search-input-form-control/map-search-input-form-control.component';
import { distinctUntilChanged, map } from 'rxjs';
import { toSignal } from '@angular/core/rxjs-interop';
import { getCenter } from 'geolib';
import { ElementRefDirective } from '@directives/element-ref/element-ref.directive';
import {
    MAP_OPTIONS,
    HEAD_OFFICE_LOCATION,
} from '@providers/google-maps-provider';

@Component({
    selector: 'app-map-form-control',
    imports: [
        MapAreaSelectFormControlComponent,
        MapPointSelectFormControlComponent,
        ElementRefDirective,
        ReactiveFormsModule,
        MapSearchInputFormControlComponent,
    ],
    providers: [
        {
            provide: NG_VALUE_ACCESSOR,
            useExisting: forwardRef(() => MapFormControlComponent),
            multi: true,
        },
    ],
    schemas: [],
    templateUrl: './map-form-control.component.html',
})
export class MapFormControlComponent implements ControlValueAccessor {
    private readonly fb = inject(FormBuilder);
    private readonly mapOptions = inject(MAP_OPTIONS);
    protected readonly headOfficeLocation = inject(HEAD_OFFICE_LOCATION);
    public readonly vm = input.required<MapFormControlVM>();
    protected readonly divSignal = signal<ElementRef<HTMLElement> | null>(null);
    private readonly targetArea = signal<Coordinates[] | null>(null);

    protected readonly mapSignal = computed(() => {
        const canvas = this.divSignal();
        if (!canvas) return null;
        return new google.maps.Map(canvas.nativeElement);
    });

    private readonly targetAreaCenter = computed<Coordinates | null>(() => {
        const area = this.targetArea();
        if (!area) return null;
        const center = getCenter(area);
        return center
            ? {
                  lat: center.latitude,
                  lng: center.longitude,
              }
            : null;
    });

    protected readonly mapCenter = computed<Coordinates>(
        () =>
            this.targetAreaCenter() ??
            this.vm().defaultCenter ??
            this.headOfficeLocation
    );

    protected readonly setMapCenterEffect = effect(
        () => {
            const map = this.mapSignal();
            if (!map) return;
            map.setCenter(this.mapCenter());
            this.setMapCenterEffect.destroy();
        },
        { manualCleanup: true }
    );

    protected mapForm = this.fb.group({
        targetArea: this.fb.control<Coordinates[] | null>(
            null,
            Validators.required
        ),
        entryPoint: this.fb.control<Coordinates | null>(
            null,
            Validators.required
        ),
    });

    private readonly setMapOptionsEffect = effect(() => {
        const map = this.mapSignal();
        if (!map) return;

        map.setOptions({
            ...this.mapOptions,
            mapTypeControlOptions: {
                position: google.maps.ControlPosition.TOP_RIGHT,
            },
        });
    });

    protected setMapBounds(bounds: google.maps.LatLngBounds | null) {
        const map = this.mapSignal();
        if (!bounds || !map) return;
        map.fitBounds(bounds);
    }

    private readonly mapFormValue: Signal<{
        targetArea: Coordinates[] | null;
        entryPoint: Coordinates | null;
    }> = toSignal(
        this.mapForm.valueChanges.pipe(
            distinctUntilChanged(),
            map((data) => {
                const { entryPoint, targetArea } = data;
                return {
                    entryPoint: entryPoint ?? null,
                    targetArea: targetArea ?? null,
                };
            })
        ),
        { initialValue: { targetArea: null, entryPoint: null } }
    );

    private readonly mapFormValueEffect = effect(() =>
        this.onChange(this.mapFormValue())
    );

    private onChange: (value: {
        targetArea: Coordinates[] | null;
        entryPoint: Coordinates | null;
    }) => void = () => {};

    private onTouched: () => void = () => {};

    public writeValue(value: {
        targetArea: Coordinates[] | null;
        entryPoint: Coordinates | null;
    }) {
        this.mapForm.setValue(value);
        this.targetArea.set(value.targetArea);
    }

    public registerOnChange(fn: () => void): void {
        this.onChange = fn;
    }

    public registerOnTouched(fn: () => void): void {
        this.onTouched = fn;
    }
}
