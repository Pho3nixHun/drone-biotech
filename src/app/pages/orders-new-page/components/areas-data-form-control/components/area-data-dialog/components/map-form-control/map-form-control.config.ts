import { EnvironmentProviders, Provider } from '@angular/core';
import { HEAD_OFFICE_LOCATION } from '@services/distance/distance.model';
import {
    GOOGLE_MAPS_LOADER_CONFIG,
    provideGoogleMapsLibraries,
} from 'src/app/shared/providers/google-maps-provider';
import { environment } from 'src/environments/environment';
import {
    POLYGON_OPTIONS,
    INFO_WINDOW_OPTIONS,
} from './components/map-area-select-form-control/map-area-select-form-control.model';
import { ENTRY_POINT_MARKER_OPTIONS } from './components/map-point-select-form-control/map-point-select-form-control.model';
import { MAP_OPTIONS } from './map-form-control.model';

export const mapFormControlConfig: (Provider | EnvironmentProviders)[] = [
    provideGoogleMapsLibraries(),
    {
        provide: POLYGON_OPTIONS,
        useValue: {
            clickable: false,
            draggable: false,
            editable: true,
            fillOpacity: 0.5,
            strokeColor: 'blue',
            fillColor: 'blue',
        },
    },
    {
        provide: MAP_OPTIONS,
        useValue: {
            mapTypeId: 'roadmap',
            disableDoubleClickZoom: true,
            clickableIcons: false,
            mapId: 'DEMO_MAP_ID',
            zoom: 14,
            streetViewControl: false,
        },
    },
    {
        provide: ENTRY_POINT_MARKER_OPTIONS,
        useValue: {
            gmpDraggable: true,
        },
    },
    {
        provide: GOOGLE_MAPS_LOADER_CONFIG,
        useValue: {
            url: 'https://maps.googleapis.com/maps/api/js?',
            key: environment.googleMapsConfig.apiKey,
            version: '3.59',
            region: 'HU',
        },
    },
    {
        provide: HEAD_OFFICE_LOCATION,
        useValue: { lat: 47.312498121576795, lng: 21.309304570654604 },
    },
    {
        provide: INFO_WINDOW_OPTIONS,
        useValue: {
            disableAutoPan: true,
            headerDisabled: true,
        },
    },
];
