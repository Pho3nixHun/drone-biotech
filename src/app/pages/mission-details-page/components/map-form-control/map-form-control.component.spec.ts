import { ComponentFixture, TestBed } from '@angular/core/testing';
import { MapFormControlComponent } from './map-form-control.component';
import { Component, DebugElement, inject } from '@angular/core';
import { FormBuilder, ReactiveFormsModule, Validators } from '@angular/forms';
import { By } from '@angular/platform-browser';
import { Coordinates } from '@stores/location/location.model';
import { provideMockMapOptions } from 'src/app/shared/providers/google-maps-provider';
import { MapAreaSelectService } from './components/map-area-select-form-control/map-area-select.service';
import { MapAreaSelectMockService } from './components/map-area-select-form-control/map-area-select-mock.service';
import { MapPointSelectService } from './components/map-point-select-form-control/map-point-select.service';
import { MapPointSelectMockService } from './components/map-point-select-form-control/map-point-select-mock.service';

@Component({
    imports: [MapFormControlComponent, ReactiveFormsModule],
    template: ` <app-map-form-control [formControl]="control" /> `,
})
class TestHostComponent {
    private readonly fb = inject(FormBuilder);
    public control = this.fb.control<{
        targetArea: Coordinates[] | null;
        entryPoint: Coordinates | null;
    }>(
        {
            targetArea: null,
            entryPoint: null,
        },
        Validators.required
    );
}
describe('MapFormControlComponent', () => {
    let fixture: ComponentFixture<TestHostComponent>;
    let component: MapFormControlComponent;
    let debugElement: DebugElement;
    let compiled: HTMLElement;

    beforeEach(async () => {
        await TestBed.configureTestingModule({
            imports: [TestHostComponent],
            providers: [provideMockMapOptions()],
        }).compileComponents();
        TestBed.overrideProvider(MapAreaSelectService, {
            useFactory: () => new MapAreaSelectMockService(),
        });
        TestBed.overrideProvider(MapPointSelectService, {
            useFactory: () => new MapPointSelectMockService(),
        });
        fixture = TestBed.createComponent(TestHostComponent);
        debugElement = fixture.debugElement.query(
            By.directive(MapFormControlComponent)
        );
        component = debugElement.componentInstance;
        compiled = debugElement.nativeElement;
    });

    // Snapshot testing
    it('should render the template correctly', () => {
        // Arrange

        // Act
        fixture.detectChanges();

        // Assert
        expect(compiled).toMatchSnapshot();
    });

    // Unit testing
    it('should set the mapSignal as a map', () => {
        // Arrange

        // Act
        fixture.detectChanges();

        // Assert
        expect(component['mapSignal']()).toBeTruthy();
    });
});
