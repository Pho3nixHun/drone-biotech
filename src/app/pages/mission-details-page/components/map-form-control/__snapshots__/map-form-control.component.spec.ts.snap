// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapFormControlComponent should render the template correctly 1`] = `
<app-map-form-control
  class="ng-untouched ng-pristine ng-valid"
  ng-reflect-form="[object Object]"
>
  <div
    appmapinitializer=""
    ng-reflect-map-options="[object Object]"
    ng-reflect-signal="[Signal: null]"
    style="height: 500px;"
  />
  <app-map-area-select-form-control
    class="ng-untouched ng-pristine ng-valid"
    ng-reflect-form="[object Object]"
    ng-reflect-map-ref="[object Object]"
  />
  <app-map-point-select-form-control
    class="ng-untouched ng-pristine ng-valid"
    ng-reflect-form="[object Object]"
    ng-reflect-map-ref="[object Object]"
  />
</app-map-form-control>
`;
